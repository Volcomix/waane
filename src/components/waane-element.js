export const html = String.raw
export const css = String.raw

export class WaaneElement extends HTMLElement {
  static get observedAttributes() {
    const propertyTypes = this.properties
    if (propertyTypes) {
      return Object.keys(propertyTypes).map(toKebabCase)
    }
  }

  constructor() {
    super()
    this._setters = {}
    this._properties = {}
    this._initShadowRoot()
    this._initProperties()
  }

  attributeChangedCallback(name) {
    this[this._setters[name]] = this[this._properties[name]]
  }

  get _template() {
    return getTemplateElement.call(this, this.tagName)
  }

  get _propertyTypes() {
    return getProperties.call(this, this.tagName)
  }

  _initShadowRoot() {
    const template = this._template
    if (!template) return

    this.attachShadow({ mode: 'open' })
    this.shadowRoot.appendChild(template.content.cloneNode(true))
  }

  _initProperties() {
    const propertyTypes = this._propertyTypes
    if (!propertyTypes) return

    Object.entries(propertyTypes).forEach(([property, type]) => {
      const attribute = toKebabCase(property)
      this._registerSetter(property, attribute)
      this._registerProperty(property, attribute, type)
    })
  }

  _registerSetter(property, attribute) {
    this._setters[attribute] = `_${property}`
  }

  _registerProperty(property, attribute, type) {
    this._properties[attribute] = property
    switch (type) {
      case Boolean:
        this._defineBooleanProperty(property, attribute)
        break
      default:
        this._defineStringProperty(property, attribute)
    }
  }

  _defineStringProperty(property, attribute) {
    Object.defineProperty(this, property, {
      get() {
        return this.getAttribute(attribute)
      },
      set(value) {
        this.setAttribute(attribute, value)
      },
    })
  }

  _defineBooleanProperty(property, attribute) {
    Object.defineProperty(this, property, {
      get() {
        return this.hasAttribute(attribute)
      },
      set(value) {
        if (value) {
          this.setAttribute(attribute, '')
        } else {
          this.removeAttribute(attribute)
        }
      },
    })
  }
}

const getTemplateElement = memoize(function () {
  const styles = this.constructor.styles
  const template = this.constructor.template
  if (!styles && !template) {
    return null
  }
  const content = []
  if (styles) {
    content.push(`<style>${styles}</style>`)
  }
  if (template) {
    content.push(template)
  }
  const templateElement = document.createElement('template')
  templateElement.innerHTML = content.join('')
  return templateElement
})

const getProperties = memoize(function () {
  return this.constructor.properties
})

const toKebabCase = memoize(function (string) {
  return string.replace(/[A-Z]/g, '-$&').toLowerCase()
})

function memoize(fn) {
  const cache = {}
  return function (argument) {
    if (argument in cache) {
      return cache[argument]
    } else {
      return (cache[argument] = fn.call(this, argument))
    }
  }
}
