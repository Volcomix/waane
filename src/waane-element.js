export function html(strings, ...substitutions) {
  const template = document.createElement('template')
  template.innerHTML = String.raw(strings, ...substitutions)
  return template
}

const templates = {}

export class WaaneElement extends HTMLElement {
  constructor() {
    super()
    this._updateMethods = {}
    this._initShadowRoot()
    this._initAttributes()
  }

  get _template() {
    const name = this.constructor.name
    let template = templates[name]
    if (!template) {
      template = this.constructor.template
      templates[name] = template
    }
    return template
  }

  _initShadowRoot() {
    this.attachShadow({ mode: 'open' })
    this.shadowRoot.appendChild(this._template.content.cloneNode(true))
  }

  _initAttributes() {
    const attributes = this.constructor.observedAttributes
    if (attributes) {
      for (const attribute of attributes) {
        this._registerUpdateMethod(attribute)
        this._registerProperty(attribute)
      }
    }
  }

  attributeChangedCallback(name, _oldValue, newValue) {
    this[this._updateMethods[name]](newValue)
  }

  _registerUpdateMethod(attribute) {
    this._updateMethods[attribute] = `_update${toPascalCase(attribute)}`
  }

  _registerProperty(attribute) {
    Object.defineProperty(this, toCamelCase(attribute), {
      get() {
        return this.getAttribute(attribute)
      },
      set(value) {
        this.setAttribute(attribute, value)
      },
    })
  }
}

function toPascalCase(string) {
  return string
    .toLowerCase()
    .replace(/(^|-)(.)/g, (_match, _p1, p2) => p2.toUpperCase())
}

function toCamelCase(string) {
  return string.toLowerCase().replace(/-(.)/g, (_match, p1) => p1.toUpperCase())
}
