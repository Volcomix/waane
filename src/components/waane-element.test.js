import { html } from './waane-element.js'

describe('html', () => {
  let template

  beforeEach(() => {
    template = html`
      <span>A template content</span>
    `
  })

  it('returns a template', () => {
    expect(template).toBeInstanceOf(HTMLTemplateElement)
  })

  it('stores the correct content inside the template', () => {
    expect(template.content).toMatchSnapshot()
  })
})

describe('WaaneElement', () => {
  it.todo('initializes the shadow root')

  it.todo('registers a _setter for each attribute')

  it.todo('registers a property for each attribute')

  it.todo('does not throw without template')

  it.todo('does not throw without observedAttributes')

  it.todo('memoizes template')

  it.todo('memoizes observedAttributes')

  it.todo('calls the associated _setter when an attribute changes')

  it.todo('reflects from property to attribute')

  it.todo('reflects from attribute to property')
})
