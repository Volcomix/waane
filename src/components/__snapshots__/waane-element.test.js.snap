// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html stores the correct content inside the template 1`] = `
<DocumentFragment>
  
      
  <span>
    A template content
  </span>
  
    
</DocumentFragment>
`;
